daemon/access.go

```diff
@@ -0,0 +1,143 @@
+	accessCancelled
+)
+
+var polkitCheckAuthorization = polkit.CheckAuthorization
+
+func checkPolkitAction(r *http.Request, ucred *ucrednet, action string) accessResult {
+	var flags polkit.CheckFlags
+	allowHeader := r.Header.Get(client.AllowInteractionHeader)
+	if allowHeader != "" {
+		if allow, err := strconv.ParseBool(allowHeader); err != nil {
+			logger.Noticef("error parsing %s header: %s", client.AllowInteractionHeader, err)
+		} else if allow {
+			flags |= polkit.CheckAllowInteraction
+		}
+	}
+	// Pass both pid and uid from the peer ucred to avoid pid race
+	if authorized, err := polkitCheckAuthorization(ucred.pid, ucred.uid, action, nil, flags); err == nil {
+		if authorized {
+			// polkit says user is authorised
+			return accessOK
+		}
+	} else if err == polkit.ErrDismissed {
+		return accessCancelled
+	} else {
+		logger.Noticef("polkit error: %s", err)
+	}
+	return accessUnauthorized
+}
+
+// accessChecker checks whether a particular request is allowed.
+//
+// An access checker will either allow a request, deny it, or return
+// accessUnknown, which indicates the decision should be delegated to
+// the next access checker.
+type accessChecker interface {
+	checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult
+}
+
+type openAccess struct{}
+
+func (ac openAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	// Forbid access from snapd-snap.socket
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
+		return accessForbidden
+	}
+
+	return accessOK
+}
+
+// authenticatedAccess allows requests from authenticated users.
+//
+// A user is considered authenticated if they provide a macaroon, are
+// the root user according to peer credentials, or granted access by
+// Polkit.
+type authenticatedAccess struct {
+	Polkit string
+}
+
+func (ac authenticatedAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	// Forbid access from snapd-snap.socket
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
+		return accessForbidden
+	}
+
+	if user != nil {
+		return accessOK
+	}
+
+	// Next checks require ucred
+	if ucred == nil {
+		return accessUnauthorized
+	}
+
+	if ucred.uid == 0 {
+		return accessOK
+	}
+
+	if ac.Polkit != "" {
+		return checkPolkitAction(r, ucred, ac.Polkit)
+	}
+
+	return accessUnauthorized
+}
+
+// rootAccess allows requests from the root uid, provided they
+// were not received on snapd-snap.socket
+type rootAccess struct{}
+
+func (ac rootAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.uid == 0 && ucred.socket != dirs.SnapSocket {
+		return accessOK
+	}
+	return accessForbidden
+}
+
+// snapAccess allows requests from the snapd-snap.socket
+type snapAccess struct{}
+
+func (ac snapAccess) checkAccess(r *http.Request, ucred *ucrednet, user *auth.UserState) accessResult {
+	if ucred != nil && ucred.socket == dirs.SnapSocket {
```

Comment:
Once again please separate the `ucred != nil` check.

On some systems snaps *may* access the system socket so perhaps we should allow both? CC @jdstrand

Reply:
It looks like the Go client library and snapd-glib will both use `snapd.socket` to call the snapctl REST API unless configured not to.  So I guess it is possible that there are some snaps with `snapd-control` that could be broken by this.  From the look of it, the existing code will let the call succeed if the snap is running as root.

So perhaps it would be best to remove the socket check all together from this checker.